crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+        anyhow::ensure!(
+            self.test_file(&self.input)?,
+            "The predicate does not consider the input Wasm file interesting."
+        );
+
+        // The smallest Wasm that passes the predicate.
+        let mut best = vec![];
+        read_file(&self.input, &mut best)?;
+
+        let new_best = |old_best: &mut Vec<u8>, new: Vec<u8>| -> Result<()> {
+            debug_assert!(
+                new.len() < old_best.len() || (new == EMPTY_WASM && *old_best == EMPTY_WASM)
+            );
+            write_file(&output, &new)?;
+            log::info!(
+                "New smallest interesting Wasm file discovered: {} bytes ({:.02}% smaller)",
+                new.len(),
+                (100.0 - (new.len() as f64 / input_size as f64 * 100.0))
+            );
+            *old_best = new;
+            Ok(())
+        };
+
+        // Next try running the predicate on an empty Wasm module.
+        //
+        // It is fairly common to accidentally write a predicate script that
+        // considers the empty module interesting, and we might as well check
+        // for it eagerly, rather than make the user wait forever until we
+        // finally to reduce the whole Wasm module to nothing.
+        if self.test_bytes(EMPTY_WASM)? {
+            new_best(&mut best, EMPTY_WASM.to_vec())?;
+            log::warn!(
+                "The predicate considers the empty Wasm module \
+                 interesting. This is usually not desired and \
+                 is a symptom of a bug in the predicate script."
+            );
+        } else {
+            // Now we perform the main search. Keep trying to find smaller and
+            // interesting variants of the current smallest interesting Wasm file
+            // until we run out of attempts and get stuck.
+
+            let mut rng = SmallRng::seed_from_u64(self.seed);
+
+            'outer: loop {
+                for _ in 0..self.attempts {
+                    let mut mutate = WasmMutate::default();
+                    mutate.reduce(true).seed(rng.gen());
+                    match mutate.run(&best) {
+                        // Can't shrink the Wasm any smaller.
+                        Err(_) => break 'outer,
+
+                        Ok(mutated_wasm) => {
+                            log::debug!("Testing variant of size {}", mutated_wasm.len());
+
+                            // This "shouldn't" happen when we set `.reduce(true)`
+                            // on `WasmMutate`, but we don't want to panic in the
+                            // middle of a long `wasm-shrink` session. Also the
+                            // `wasm-shrink` user isn't in a position to do anything
+                            // useful about this, even if we had a warning message,
+                            // so just ignore it.
+                            if mutated_wasm.len() >= best.len() {
+                                continue;
+                            }
```

Comment:
This seems like something it'd be good to have a `debug_assert!` for which could be tested while fuzzing and on our own CI?

Reply:
I think we would trip the assertion right now, so I'd rather do this in a follow up where we also fix `wasm-mutate`'s `reduce(true)` at the same.

Reply:
Ah ok, can you write that in the comment as a TODO? (to replace with `debug_assert!` when `wasm-mutate` works better)